src/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -135,30 +135,133 @@ public static unsafe Array CreateInstance(Type elementType, int[] lengths, int[]
         // Copies length elements from sourceArray, starting at index 0, to
         // destinationArray, starting at index 0.
         //
-        public static void Copy(Array sourceArray, Array destinationArray, int length)
+        public static unsafe void Copy(Array sourceArray, Array destinationArray, int length)
         {
             if (sourceArray == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.sourceArray);
             if (destinationArray == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.destinationArray);
 
-            Copy(sourceArray, sourceArray.GetLowerBound(0), destinationArray, destinationArray.GetLowerBound(0), length, false);
+            MethodTable* pMT = RuntimeHelpers.GetMethodTable(sourceArray);
+            if (pMT == RuntimeHelpers.GetMethodTable(destinationArray) &&
+                !pMT->IsMultiDimensionalArray &&
+                (uint)length <= (uint)sourceArray.Length &&
+                (uint)length <= (uint)destinationArray.Length)
+            {
+                nuint byteCount = (uint)length * (nuint)pMT->ComponentSize;
+                ref byte src = ref sourceArray.GetRawSzArrayData();
+                ref byte dst = ref destinationArray.GetRawSzArrayData();
+
+                if (pMT->ContainsGCPointers)
+                    Buffer.BulkMoveWithWriteBarrier(ref dst, ref src, byteCount);
```

Comment:
>BulkMoveWithWriteBarrier [](start = 27, length = 24)

`BulkMoveWithWriteBarrier` is still a potentially lengthy FCALL. Not the point of this change, just looked through its implementation.

Reply:
Yep, BulkMoveWithWriteBarrier  is on my TODO list.